src/main/java/org/opensearch/security/transport/SecurityInterceptor.java

```diff
@@ -158,6 +162,14 @@ public <T extends TransportResponse> void sendRequestDecorate(
         final boolean isDebugEnabled = log.isDebugEnabled();
 
         final boolean isSameNodeRequest = localNode != null && localNode.equals(connection.getNode());
+        final Set<String> requestHeadersToCopy = new HashSet<>();
+        if (getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_REQUEST_HEADERS) != null) {
+            Collections.addAll(
+                requestHeadersToCopy,
+                getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_REQUEST_HEADERS).split(",")
+            );
+            requestHeadersToCopy.remove(Task.X_OPAQUE_ID); // Special case where this header is preserved during stashContext.
```

Comment:
The changes in this file are necessary because currently only `X_OPAQUE_ID` is not lost when stashing here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L157-L163

Ideally, this is solved another way through the core. Either:

1. Change X_OPAQUE_ID to a persistent header and remove the logic in `stashContext`. persistentHeaders are a special category of headers that are not stashable.
2. Update the logic in ThreadContext to not stash any of the headers from https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/ActionModule.java#L581-L586

Reply:
IMO, option 1 is better.